### PR TITLE
INT-4115: (S)FtpPersistentFileFilter by default

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4115

Apply `(S)FtpPersistentAcceptOnceFileListFilter` for the `(S)FtpInboundFileSynchronizer` by default to avoid cases to sync the same remote files to the local directory again.
Especially when `localFileName` strategy is applied and we end up with new local files, but with the same remote content

### CODE REVIEW

**Comment by garyrussell** [comment]
I am a bit confused. INT-4115 is about the `AbstractRemoteFile__Streaming__MessageSource` and this PR does not address that.

Further, when configuring with XML, we have pattern (and filter) attributes.

Don't we need to combine this into a composite (or chain) filter when configured with XML?

**Comment by artembilan** [comment]
Oh! Really?
I have missed that 😄 

But is it still applied for the `AbstractInboundFileSynchronizer` impl? Isn't it?

Re, XML parsing we have this:
```
parserContext.getReaderContext().error("at most one of 'filename-pattern', " +
		"'filename-regex', 'filter' or 'filter-expression' is allowed on remote file inbound adapter",
						element);
```
So, I think no, we shouldn't add more complexity there.
More over the local part is only like this:
```
IntegrationNamespaceUtils.setReferenceIfAttributeDefined(messageSourceBuilder, element, "local-filter");
IntegrationNamespaceUtils.setValueIfAttributeDefined(messageSourceBuilder, element, "local-directory");
```
So, we still have to build composition manually here.

File: spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java [hunk 1]
```diff
@@ -165,6 +168,10 @@ public void setRemoteDirectoryExpression(Expression remoteDirectoryExpression) {
 	 * @param filter the file list filter.
 	 */
 	public void setFilter(FileListFilter<F> filter) {
+		doSetFilter(filter);
+	}
+
+	public void doSetFilter(FileListFilter<F> filter) {
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java]: Did you intend to make this `final` ?

File: spring-integration-file/src/main/java/org/springframework/integration/file/remote/AbstractRemoteFileStreamingMessageSource.java [hunk 2]
```diff
@@ -109,6 +111,10 @@ public void setRemoteFileSeparator(String remoteFileSeparator) {
 	 * @param filter the file list filter.
 	 */
 	public void setFilter(FileListFilter<F> filter) {
+		doSetFilter(filter);
+	}
+
+	public void doSetFilter(FileListFilter<F> filter) {
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/AbstractRemoteFileStreamingMessageSource.java]: `final` ?

**Review by garyrussell — CHANGES_REQUESTED** [review]
I am still a bit concerned by this; I would guess that most people apply a pattern filter of some kind, which effectively removes this new filter. I would say we should always combine it in a composite filter. Otherwise, this change won't be applied in most cases.

File: spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java (hunk 1 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java]: Hm... Seems for me I was going to make them `private`.
Does it make sense?

Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java]: but these are in the abstract class and called from the subclasses.

`protected final ...` ??

Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java]: Doh. Yeah, really `protected final`.
That was already some time ago when this PR has been fresh in my mind. 😄

**Comment by artembilan** [comment]
> I am still a bit concerned by this; I would guess that most people apply a pattern filter of some kind, which effectively removes this new filter. I would say we should always combine it in a composite filter. Otherwise, this change won't be applied in most cases.

OK. I think I see your point. 
And the idea is to use something like `FileListFilterFactoryBean` what we have from the `FileInboundChannelAdapterParser`.

Not sure that we always have to add it into the composite, but at least we really can do that when `regexp` or `pattern` is in use. For any other custom case there is always `CompositeFileListFilter` to configure externally and inject.

That is already like an improvement, next steps, to the current fix.
I think it is still valid to have it like thing by default.

**Comment by garyrussell** [comment]
>I think it is still valid to have it like thing by default.

I agree; my concern is that it won't actually be a default in most cases, because people generally use pattern or regex.

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java [hunk 3]
```diff
@@ -49,7 +51,8 @@ protected final BeanMetadataElement parseSource(Element element, ParserContext p
 
 		// configure the InboundFileSynchronizer properties
 		BeanDefinition expressionDef = IntegrationNamespaceUtils.createExpressionDefinitionFromValueOrExpression(
-				"remote-directory", "remote-directory-expression", parserContext, element, false);
+				"remote-directory", "remote-directory-expression", parserContext,
+				element, false);
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: I am just wondering what drove this change - unnecessary reformatting causes loss of blame.

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java [hunk 4]
```diff
@@ -118,7 +124,7 @@ static void configureFilter(BeanDefinitionBuilder synchronizerBuilder, Element e
 			}
 			if (count != 1) {
 				parserContext.getReaderContext().error("at most one of 'filename-pattern', " +
-						"'filename-regex', 'filter' or 'filter-expression' is allowed on remote file inbound adapter",
+								"'filename-regex', 'filter' or 'filter-expression' is allowed on remote file inbound adapter",
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: Bad indent - pushed past 120.

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java [hunk 5]
```diff
@@ -132,16 +138,44 @@ else if (hasFilterExpression) {
 				synchronizerBuilder.addPropertyValue("filter", expressionFilterBeanDefinition);
 			}
 			else if (hasFileNamePattern) {
-				BeanDefinitionBuilder filterBuilder = BeanDefinitionBuilder.genericBeanDefinition(patternClass);
-				filterBuilder.addConstructorArgValue(fileNamePattern);
-				synchronizerBuilder.addPropertyValue("filter", filterBuilder.getBeanDefinition());
+				BeanDefinition patterFilter =
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: `s/patter/pattern/`

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java [hunk 6]
```diff
@@ -94,4 +99,6 @@ protected final BeanMetadataElement parseSource(Element element, ParserContext p
 
 	protected abstract Class<? extends FileListFilter<?>> getRegexPatternFileListFilterClass();
 
+	protected abstract Class<? extends AbstractPersistentAcceptOnceFileListFilter<?>> getPersistentAcceptOnceFileListFilterClass();
```
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: Perhaps return null by default? (Allow subclasses, e.g. AWS to not provide a filter).

Maybe not - since this is a major release, I guess it's ok.

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java (hunk 3 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: The line exceeds 120 symbols.

I feel like there might be a lot of code which we won't touch in the future at all.
But its problem that it might be inconsistent with our code style.
So, if we request 120 line length we should prove that with all our code.
Otherwise we may meet some nasty commentators that we don't do that in our own code. 😄 

I know that you don't like such a code style reformatting for nothing, but sometimes doing that via IDEA tool I meet some interesting violations, like extra whitespaces in the code line, missed blank lines around class members etc...

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java (hunk 4 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: I wish my IDEA has the same formatting settings as your STS, but I really don't know how to configure this indent settings....

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java (hunk 3 above)
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: ?? The previous line length was 98

I have no issue when the reformat is to comply with our style but this doesn't fit that rule.

Hence my question.

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java (hunk 4 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: @snicoll ,

any advice for IntelliJ IDEA?
Thanks

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java (hunk 3 above)
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: And the line above is clearly longer (113).

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java (hunk 4 above)
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: But the extra indent pushed it past 120 - why did you even touch this line? - do you have some "auto-format on save"

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java (hunk 3 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: Doh...
In new IDEA it looks like:
idea_120

The visual perception deceived me... 😢 

My bad, sorry.
Will be much careful in the future.
Reverting...

Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java]: Sometimes these IDEs try to be too smart 😄 - if it adds metadata, it should push the margin out appropriately for those lines.

File: spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java (hunk 4 above)
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: I used the reformatting tool manually (Ctrl+Alt+L) and it is applied for the whole code.

Again: I don't know how to make it only for the touch code and don't know how to not miss violations in other lines.

I guess if I have the proper setting for these kind of indents, even if I perform that reformatting, this line of code wouldn't be touch and we wouldn't have such a long discussion....

Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: In STS, Ctrl-Shift-F (with nothing selected) reformats the whole file, but if you just select the lines you are interested in it only reformats the selected lines. I try to avoid Ctrl-Shift-F without making a selection first.

I wouldn't mind (although as you know I dislike unnecessary formatting), but in this case, it pushed it past 120, which is bad.

Comment (snicoll) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: we have the same issue in Spring Boot. @philwebb was working on a generic formatter solution but this had to be put on hold to work on other things. It's current ad-hoc at the moment and I don't think you can't configure STS and IJ to match perfectly (especially with indent, the algorithm doesn't match for everything).

Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/config/FileParserUtils.java]: Thank you @snicoll for sharing your pain as  well 😄 

I think we with @garyrussell will work out some formal rules to follow with formatting.

**Comment by garyrussell** [comment]
Merged as 10ce68d3e33e38ecaf109b866fedb2fd0408ab3b
